packages/vitest-plugin-rsc/src/testing-library-client.tsx

```diff
@@ -0,0 +1,89 @@
+import * as React from "react";
+import * as ReactDOMClient from "react-dom/client";
+import * as ReactClient from "@vitejs/plugin-rsc/react/browser";
+import type { RenderConfiguration } from "../dist/testing-library";
+
+export type RscPayload = {
+  root: React.ReactNode;
+  returnValue?: unknown;
+};
+
+export type TestingLibraryClientRoot = Awaited<
+  ReturnType<typeof createTestingLibraryClientRoot>
+>;
+
+export type FetchRsc = (actionRequest?: {
+  id: string;
+  reply: string | FormData;
+}) => Promise<ReadableStream<Uint8Array>>;
+
+export async function createTestingLibraryClientRoot(options: {
+  container: HTMLElement;
+  config: RenderConfiguration;
+  fetchRsc: FetchRsc;
+}) {
+  let setPayload: (v: RscPayload) => void;
+
+  const initialPayload = await ReactClient.createFromReadableStream<RscPayload>(
+    await options.fetchRsc(),
+  );
+
+  function BrowserRoot() {
+    const [payload, setPayload_] = React.useState(initialPayload);
+
+    React.useEffect(() => {
+      setPayload = (v) => React.startTransition(() => setPayload_(v));
+    }, [setPayload_]);
+
+    return payload.root;
+  }
+
+  ReactClient.setServerCallback(async (id, args) => {
+    const temporaryReferences = ReactClient.createTemporaryReferenceSet();
+    const reply = await ReactClient.encodeReply(args, { temporaryReferences });
+    const payload = await ReactClient.createFromReadableStream<RscPayload>(
+      await options.fetchRsc({ id, reply }),
+      { temporaryReferences },
+    );
+    setPayload(payload);
+    return payload.returnValue;
+  });
```

Comment:
I think server callback should be implemented through `encodeReply/decodeReply`. Even though we can technically pass `args` directly since they are on the same browser, this probably matches better with an actual behavior.

Reply:
Nice improvement!